UInput.cpp

```diff
@@ -60,9 +60,20 @@ int fpc_uinput_create(fpc_uinput_t *uinput)
 
     rc = ioctl(fd, UI_DEV_SETUP, &usetup);
     if (rc < 0) {
-        ALOGE("Failed to setup uinput device! rc=%d: %s", rc, strerror(errno));
-        close(fd);
-        return rc;
+        ALOGW("Failed to setup uinput device! rc=%d: %s, falling back to write.", rc, strerror(errno));
+
+        struct uinput_user_dev usetup_legacy;
+        memset(&usetup_legacy, 0, sizeof(usetup_legacy));
+        usetup_legacy.id.bustype = BUS_VIRTUAL;
+        strcpy(usetup_legacy.name, usetup.name);
+
+        rc = write(fd, &usetup_legacy, sizeof(usetup_legacy));
+
+        if (rc < 0) {
+            ALOGE("Failed to setup uinput device! rc=%d: %s", rc, strerror(errno));
+            close(fd);
+            return rc;
+        }
```

Comment:
I personally always check whether the full object was written, such as in `fpc_write_input_event`. Though that case should really be improved to print the strerror too, like:

```suggestion
        if (rc < 0) {
            ALOGE("Failed to write legacy uinput setup! rc=%d: %s", rc, strerror(errno));
            close(fd);
            return rc;
        } else if (rc != sizeof(usetup_legacy)) {
            ALOGE("Didn't write full usetup_legacy, only %d/%zu bytes!", rc, sizeof(usetup_legacy));
            /*
            The kernel _should_ always accept the full object, and a loop to write the whole thing is overkill too. shall we just close it and return an error?
            close(fd);
            return -EBUSY;
            */
        }
```
I don't expect this to ever happen though. But better be safe than sorry.